src/attribute.cpp

```diff
@@ -27,6 +43,175 @@ Integer::Integer(uint8_t type, const uint8_t* data, size_t size)
+        : Attribute(type)
+{
+    if (size > 128)
+        throw std::runtime_error("Invalid encrypted attribute size. Should be max 128 bytes, actual size is " + std::to_string(size));
+
+    std::vector<uint8_t> value(size);
+
+    size_t j = 16;
+    while (j <= size)
+    {
+        std::vector<uint8_t> buffer(16 + secret.length());
+
+        for (size_t i = 0; i < secret.length(); ++i)
+            buffer[i] = secret[i];
+
+        if (j == 16)
+        {
+            for (size_t i = 0; i < 16; ++i)
+                buffer[i + secret.length()] = auth[i];
+        }
+        else
+        {
+            for (size_t i = 0; i < 16; ++i)
+                buffer[i + secret.length()] = data[i + j - 32];
+        }
+        std::array<uint8_t, 16> md;
+
+        MD5(buffer.data(), buffer.size(), md.data());
+
+        for (size_t i = 0; i < 16; ++i)
+            value.push_back(data[i + j - 16] ^ md[i]);
+
+        j += 16;
+    }
+    m_value.assign(value.begin(), value.end());
+}
+
+Encrypted::Encrypted(uint8_t type, std::string password)
+    : Attribute(type),
+      m_value(password)
+{
+}
+
+std::vector<uint8_t> Encrypted::toVector(const std::string& secret, std::array<uint8_t, 16> auth) const
+{
+    if (m_value.length() > 128)
+        throw std::runtime_error("Invalid encrypted attribute size. Should be max 128 bytes, actual size is " + std::to_string(m_value.length()));
+
+    std::string value = m_value;
+
+    if (value.length() % 16 != 0)
+    {
+        for (size_t i = 0; i < (16 - m_value.length() % 16); ++i)
+            value.push_back(0);
+    }
+
+    std::vector<uint8_t> buffer(16 + secret.length());
+
+    for (size_t i = 0; i < secret.length(); ++i)
+        buffer[i] = secret[i];
+    for (size_t i = 0; i < 16; ++i)
+        buffer[i + secret.length()] = auth[i];
+
+    std::array<uint8_t, 16> md;
+    MD5(buffer.data(), buffer.size(), md.data());
+
+    std::vector<uint8_t> attribute;
+    std::vector<uint8_t> result(16);
+    for (size_t j = 0; j < value.length() / 16; ++j)
+    {
+        for (size_t i = 0; i < 16; ++i)
+        {
+            result[i] = value[i + j * 16] ^ md[i];
+            attribute.push_back(result[i]);
+        }
+        if (j != value.length() - 1)
+        {
+            for (size_t i = 0; i < 16; ++i)
+                buffer[i + secret.length()] = result[i];
+
+            MD5(buffer.data(), buffer.size(), md.data());
+        }
+    }
+
+    auto it = attribute.begin();
+    it = attribute.insert(it, attribute.size() + 2);
+    it = attribute.insert(it, type());
+    return attribute;
+}
+
+std::string byteToHex(uint8_t number)
+{
+    const std::string digits = "0123456789ABCDEF";
+    std::string hex;
+    hex.append(1, digits[number / 16]);
+    hex.append(1, digits[number % 16]);
+    return hex;
+}
+
+Bytes::Bytes(uint8_t type, const uint8_t* data, size_t size)
+        : Attribute(type)
+{
+    for (size_t i = 0; i < size; ++i)
+        m_value.push_back(data[i]);
```

Comment:
It is always wise to allocate vector to its size, when it is known. So, a better approach is:
```c++
Bytes::Bytes(uint8_t type, const uint8_t* data, size_t size)
    : Attribute(type),
      m_value(size)
{
    for (size_t i = 0; i < size; ++i)
        m_value[i] = data[i];
}
```